### PR TITLE
Proposal Linking Backend implementation & tests

### PR DESCRIPTION
**Pull request by ianrowan**
## Link to Issue
Closes: #3541  

## Description of Changes
- Database migration to add JSONB links column to Threads + consolidate existing links(snapshot, threads, chainEntityMeta) into new column for existing records
- New routes
    - `POST /linking/getLinks`: get bi-directional links for threads <-> (threads, snapshot, proposals)
    - `POST /linking/addThreadLinks`: add Links to a existing thread_id
    - `DELETE /linking/deleteLinks`: delete links from Thread Id
- API integration tests 

## Test Plan
- Mocked a jwt in postman and tried all configs 
- Built and confirmed all integration tests succeed with full coverage

## Deployment Plan
1. Currently the migration will need to run + frontend update to use the routes on both read/write. This migration does not delete old data to ensure easy rollback. Please confirm if this is the correct approach.

## Other Considerations
- Proposals are stored generically, so FE will need to know how to build a proposal link based on the chain(ie ChainEntities for evm chains, ? for cosmos chains) 
- In the migration, links in the `ChainEntityMeta` table were consolidated as links of type `proposal` with `id: ce_id` please advise if this is how we want to approach this migration

### CODE REVIEW

File: packages/commonwealth/server/migrations/20230421205344-add_thread_link.js [hunk 1]
```diff
@@ -2,10 +2,68 @@
 
 module.exports = {
   up: async (queryInterface, Sequelize) => {
-    return queryInterface.addColumn('Threads', 'links', {
+    await queryInterface.addColumn('Threads', 'links', {
```
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20230421205344-add_thread_link.js:5]: Migration should be transactionalized. Note that this makes the loop here tricky, as we will need to lock the entire table for the duration of the migration. I wonder if there's a cleaner way we can make these updates...

File: packages/commonwealth/server/routes/linking/addThreadLinks.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import { AppError } from 'common-common/src/errors';
+import type { NextFunction, Request, Response } from 'express';
+import { Link, LinkSource } from '../../models/thread';
+import type { DB } from '../../models';
+import { Errors, isAuthorOrAdmin } from '../../util/linkingValidationHelper';
+
+const addThreadLink = async (
+  models: DB,
+  req: Request,
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/linking/addThreadLinks.ts:9]: Can we used `TypedRequest[Body|Query]` and `TypedResponse` here / throughout?

**Review by jnaviask — CHANGES_REQUESTED** [review]
Generally happy with this aside from a few small comments. Would like to investigate the migration, specifically it needs to be within a sequelize transaction + then I worry about migration performance / whether we may encounter deadlocks.

Overall, though, looks great.

**Comment by jnaviask** [comment]
@ianrowan feel free to take over frack and test out the performance of the migration. Hopefully is under 5s.

Also, are the unit test failures expected?

**Comment by ianrowan** [comment]
@jnaviask 
image
Looks like its under 5s 

And as for unit tests - the failure is an un-related file/test related to react

File: packages/commonwealth/server/migrations/20230421205344-add_thread_link.js [hunk 3]
```diff
@@ -2,10 +2,84 @@
...
-    });
+    const transaction = await queryInterface.sequelize.transaction();
+    try {
+      await queryInterface.addColumn('Threads', 'links', {
+        type: Sequelize.JSONB,
+        allowNull: true,
+        transaction,
+      });
+
+      const [threads, metadata] = await queryInterface.sequelize.query(
```
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20230421205344-add_thread_link.js:13]: We should have these read queries as part of the transaction as well.

**Review by jnaviask — CHANGES_REQUESTED** [review]
One more small fix then we can merge this.

**Review by jnaviask — APPROVED** [review]
Will merge to feature then rereview before merge to master

**Pull request merged**